src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp

```diff
@@ -145,69 +105,57 @@ CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t
+    CHIP_ERROR err         = CHIP_ERROR_INVALID_ARGUMENT;
+    bool_t putKey          = false;
+    uint8_t pdmIdKvs       = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
+    uint16_t pdmInternalId = 0;
+    uint16_t freeKeyId;
+    uint8_t keyId;
 
-            if (iter == g_key_ids_set.end())
-            {
-                if (g_key_ids_set.size() >= kMaxNumberOfKeys)
-                {
-                    return CHIP_ERROR_NO_MEMORY;
-                }
+    VerifyOrExit((key != NULL) && (value != NULL), err = CHIP_ERROR_INVALID_ARGUMENT);
 
-                g_key_ids_set.insert(key_id);
+    keyId = GetStringKeyId(key, &freeKeyId);
 
-                put_key = true;
-                break;
-            }
-        }
+    // Key allready exists
+    if (keyId < kMaxNumberOfKeys)
+    {
+        // Update just the value in this case
+        putKey = false;
     }
-    else /* overwrite key */
+    else
     {
-        put_key = true;
-        key_id  = it->second;
+        // Need to write both the value and the string key
+        putKey = true;
+        keyId  = freeKeyId;
     }
 
-    if (put_key)
-    {
-        pdm_internal_id = chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id);
-        ChipLogProgress(DeviceLayer, "KVS, save in flash the value of the Matter key [%s] with PDM id: %i", key, pdm_internal_id);
+    pdmInternalId = chip::DeviceLayer::Internal::K32WConfigKey(pdmIdKvs, keyId);
+    ChipLogProgress(DeviceLayer, "KVS, save in flash the value of the Matter key [%s] with PDM id: %i", key, pdmInternalId);
 
-        g_kvs_map.insert(std::make_pair(std::string(key), key_id));
-        err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueBin(pdm_internal_id, (uint8_t *) value, value_size);
+    err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueBin(pdmInternalId, (uint8_t *) value, value_size);
 
-        /* save the 'key' in flash such that it can be retrieved later on */
-        if (err == CHIP_NO_ERROR)
+    /* save the 'key' in flash such that it can be retrieved later on */
+    if (err == CHIP_NO_ERROR)
+    {
+        if (true == putKey)
         {
-            pdm_internal_id = chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id + kMaxNumberOfKeys);
-            ChipLogProgress(DeviceLayer, "KVS, save in flash the Matter key [%s] with PDM id: %i", key, pdm_internal_id);
+            pdmInternalId = chip::DeviceLayer::Internal::K32WConfigKey(pdmIdKvs, keyId + kMaxNumberOfKeys);
+            ChipLogProgress(DeviceLayer, "KVS, save in flash the Matter key [%s] with PDM id: %i", key, pdmInternalId);
 
             /* TODO (MATTER-132): do we need to make sure that "key" is NULL-terminated? */
-            err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueStr(pdm_internal_id, key, strlen(key) + 1);
+            err = chip::DeviceLayer::Internal::K32WConfig::WriteConfigValueStr(pdmInternalId, key, strlen(key) + 1);
```

Comment:
If I understand the PR description, this is using an in-memory cache in the matter event loop, with write through to flash happening in a separate thread.

If so, we need some thread synchronization between the two contexts.  What provides this?

Apologies if I'm misunderstanding the desgin.